src/codec/avro/mod.rs

```diff
@@ -0,0 +1,125 @@
+use std::error::Error;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use serde_avro_fast::ser::SerializerConfig;
+use serde_avro_fast::Schema;
+
+mod registry;
+pub use registry::AvroRegistry;
+
+use super::Codec;
+
+#[derive(Debug)]
+pub struct SchemaDefinition {
+    pub schema: Schema,
+    pub id: u32,
+}
+
+#[async_trait]
+pub trait SchemaRegistry {
+    type Error: Error + Send + Sync + 'static;
+
+    async fn schema(&self, topic: &str, id: u32) -> Result<Arc<SchemaDefinition>, Self::Error>;
+
+    async fn default_schema(&self, topic: &str) -> Result<Arc<SchemaDefinition>, Self::Error>;
+}
+
+#[derive(Debug)]
+pub struct Avro<R: SchemaRegistry> {
+    registry: Arc<R>,
+    topic: String,
+}
+
+impl<R> Clone for Avro<R>
+where
+    R: SchemaRegistry,
+{
+    fn clone(&self) -> Self {
+        Self {
+            registry: self.registry.clone(),
```

Comment:
* Consider wrapping registry and topic into `Arc` to avoid cloning strings
* I personally prefer `Arc::clone(&x)` over `x.clone()`, because of the default reasoning "`clone()` equals allocation".

Reply:
i wrapped it on a Codec level but it's still cloned later. Generally i don't like the idea topic coming from configuration level but i didn't want to change signature of coded and extend it with topic name which is a message bus specific thing :(